### PR TITLE
Issue #27: Importing a diagram from URL doesn't work properly

### PR DESCRIPTION
**Pull request by arcilli**
Used DiagramSheet for sharing `getDiagramXMLFromUrl`, which is used by both Edit & View Sheet.
Also, import from URL is now working in the same manner in which a diagram is decoded & displayed in ViewSheet.

### CODE REVIEW

File: src/main/resources/Diagram/DiagramEditSheet.xml [hunk 1]
```diff
@@ -204,6 +204,7 @@ require.config({
     'spin': diagramEditorBasePath + 'js/spin/spin.min',
     'jszip': diagramEditorBasePath + 'js/jszip/jszip.min',
     'draw.io': diagramEditorBasePath + 'js/draw.io.min',
+    'diagram-utils' : new XWiki.Document('DiagramSheet', 'Diagram', '').getURL('jsx', 'param=1')
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramEditSheet.xml]:
```suggestion
    'diagram-utils': new XWiki.Document('DiagramSheet', 'Diagram').getURL('jsx', 'param=1')
```
You can simply omit the wiki parameter. No need to pass the empty string.

File: src/main/resources/Diagram/DiagramEditSheet.xml [hunk 2]
```diff
@@ -215,6 +216,11 @@ require.config({
     'draw.io': {
       deps: ['mxgraph-editor', 'base64', 'pako-global', 'spin-global', 'jszip']
     }
+  },
+  map : {
+    'diagram-utils' : {
+      'mxgraph-common' : 'mxgraph-editor'
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramEditSheet.xml]:
```suggestion
      'mxgraph-common': 'mxgraph-editor'
```
I see you are used to leave a space before ``:``. So far our code style is to leave a space only after ``:`` (same as with ``,``).

File: src/main/resources/Diagram/DiagramEditSheet.xml [hunk 3]
```diff
@@ -767,6 +774,24 @@ define('diagramStore', ['jquery', 'xwiki-meta', 'xwiki-utils', 'draw.io', 'xwiki
...
   }
 
+  var originalLoadUrl = EditorUi.prototype.loadUrl;
+  EditorUi.prototype.loadUrl = function(url, success, error, forceBinary, retry, dataUriPrefix) {
+    let proxyIndex = url.indexOf(PROXY_URL);
+    // Use indexOf since URLSearchParams is not supported by IE.
+    let urlParamIndex = url.indexOf('url=');
+    if (proxyIndex &gt; -1 &amp;&amp; urlParamIndex &gt; -1) {
+      // Get the value of url parameter
+      let originalUrl = unescape(url.substring(urlParamIndex+4));
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramEditSheet.xml]: You're making 2 assumptions here:

* there's no other parameter in the query string after the ``url`` parameter
* there's no document fragment identifier at the end of the URL

It may be the case but it's not clean and it makes the code fragile.

Also, on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/unescape I see this note:

> Note: Do not use unescape to decode URIs, use decodeURI instead.

Finally, I don't understand why you need to decode twice the value of the ``url`` parameter. Calling **only** ``decodeURIComponent`` should be enough I think.

File: src/main/resources/Diagram/DiagramSheet.xml [hunk 4]
```diff
@@ -48,4 +48,122 @@
...
+      <defaultWeb/>
+      <nameField/>
+      <validationScript/>
+      <cache>
+        <cache>0</cache>
+        <defaultValue>long</defaultValue>
+        <disabled>0</disabled>
+        <displayType>select</displayType>
+        <freeText>forbidden</freeText>
+        <largeStorage>0</largeStorage>
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramSheet.xml]: This is not good. See https://github.com/xwikisas/application-licensing/issues/67#issuecomment-523871959 .

File: src/main/resources/Diagram/DiagramSheet.xml [hunk 5]
```diff
@@ -48,4 +48,122 @@
...
+  return {
+    getDiagramXMLFromUrl : getDiagramXMLFromUrl
+  };
+});</code>
+    </property>
+    <property>
+      <name>Diagram utils</name>
+    </property>
+    <property>
+      <parse>1</parse>
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramSheet.xml]: There's no Velocity code in the JSX so there's no need to parse it.

File: src/main/resources/Diagram/DiagramSheet.xml [hunk 6]
```diff
@@ -48,4 +48,122 @@
...
+    </class>
+    <property>
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>/*
+ * Used in view &amp; edit sheets
+ */
+define ('diagram-utils', ['mxgraph-common'], function() {
+  function getDiagramXMLFromUrl(url) {
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramSheet.xml]: The function name is not nice for 2 reasons:
* it's not consistent: it should be either ``getDiagramXMLFromURL`` or ``getDiagramXmlFromUrl``
* The code style we've applied so far is to use upper case for abbreviations, so ``getDiagramXMLFromURL``

File: src/main/resources/Diagram/DiagramViewSheet.xml [hunk 7]
```diff
@@ -177,7 +177,8 @@ require.config({
     'sanitizer': mxGraphViewerBasePath + 'sanitizer/sanitizer.min',
     'mxgraph-viewer': mxGraphViewerBasePath + 'mxGraphViewer.min',
     'draw.io.viewer': diagramViewerBasePath + 'js/draw.io.viewer.min',
-    'pako': diagramViewerBasePath + 'js/deflate/pako.min'
+    'pako': diagramViewerBasePath + 'js/deflate/pako.min',
+    'diagram-utils' : new XWiki.Document('DiagramSheet', 'Diagram', '').getURL('jsx', 'param=1')
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramViewSheet.xml]:
```suggestion
    'diagram-utils': new XWiki.Document('DiagramSheet', 'Diagram').getURL('jsx', 'param=1')
```

File: src/main/resources/Diagram/DiagramViewSheet.xml [hunk 8]
```diff
@@ -325,20 +332,9 @@ define('pako-global', ['pako'], function(pako) {
...
-        let rawDiagram = graph.decompress(decodedData);
-        return rawDiagram;
-    }
-    return null;
-  };
-
-  function getDiagramXML(container) {
-    return getDiagramXMLFromURL() || container.data('model');
+  function getDiagramXML(url, container) {
+    let unescapedUrl = unescape(url);
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramViewSheet.xml]:
```suggestion
    let unescapedUrl = decodeURI(url);
```
As per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/unescape

File: src/main/resources/Diagram/DiagramEditSheet.xml [hunk 9]
```diff
@@ -767,6 +774,20 @@ define('diagramStore', ['jquery', 'xwiki-meta', 'xwiki-utils', 'draw.io', 'xwiki
     oldImportFiles.apply(this, newArgs);
   }
 
+  var originalLoadUrl = EditorUi.prototype.loadUrl;
+  EditorUi.prototype.loadUrl = function(url, success, error, forceBinary, retry, dataUriPrefix) {
+    let exportedUrl = diagramUtils.getParameterValueFromURL('url', url);
+    if (exportedUrl) {
+      let decodedUrl = decodeURIComponent(exportedUrl);
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramEditSheet.xml]: Why do you need to decode the parameter value? ``getParameterValueFromURL`` does it already.

File: src/main/resources/Diagram/DiagramSheet.xml [hunk 10]
```diff
@@ -48,4 +48,138 @@
...
+      // Exclude first 2 letters from hash (#R).
+      let decodedData = hash.substring(hashIndex+2);
+      let graph = new Graph();
+      let rawDiagram = graph.decompress(decodedData);
+      return rawDiagram;
+    }
+    return null;
+  };
+
+  $.urlParam = function(parameter, url) {
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramSheet.xml]: I don't think it's a good idea to extend the jQuery object. Why not define a local function?

File: src/main/resources/Diagram/DiagramSheet.xml [hunk 11]
```diff
@@ -48,4 +48,138 @@
...
+       return null;
+    }
+    return decodeURIComponent(results[1]);
+  }
+
+  function getParameterValueFromURL(parameter, url) {
+    if (typeof URLSearchParams === 'function') {
+      let rawUrl = new URL (url, window.location.href);
+      let searchParams = new URLSearchParams(rawUrl.search);
+      if (searchParams.has(parameter)) {
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramSheet.xml]: Do you really need to test if the parameter is present? Why not return directly the value, which I assume is null if the parameter is not specified.

File: src/main/resources/Diagram/DiagramViewSheet.xml [hunk 12]
```diff
@@ -325,20 +332,9 @@ define('pako-global', ['pako'], function(pako) {
...
-        let graph = new Graph();
-        let rawDiagram = graph.decompress(decodedData);
-        return rawDiagram;
-    }
-    return null;
-  };
-
-  function getDiagramXML(container) {
-    return getDiagramXMLFromURL() || container.data('model');
+  function getDiagramXML(url, container) {
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramViewSheet.xml]: You're calling this function with ``window.location.hash`` above so the first parameter seems to be a hash not a full URL. Which part is right?

File: src/main/resources/Diagram/DiagramViewSheet.xml [hunk 13]
```diff
@@ -325,20 +332,9 @@ define('pako-global', ['pako'], function(pako) {
...
+    let decodedUrl = decodeURIComponent(url);
+    return diagramUtils.getParameterValueFromURL(decodedUrl) || container.data('model');
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramViewSheet.xml]: ``getParameterValueFromURL`` expects two parameters so something is not right here.

Comment (arcilli) [reply src/main/resources/Diagram/DiagramViewSheet.xml]: Yes, `getDiagramXMLFromURL` should be called instead of `getParameterValueFromURL`. Probably I didn't catch it because of caching.

File: src/main/resources/Diagram/DiagramEditSheet.xml [hunk 14]
```diff
@@ -767,6 +774,19 @@ define('diagramStore', ['jquery', 'xwiki-meta', 'xwiki-utils', 'draw.io', 'xwiki
...
   }
 
+  var originalLoadUrl = EditorUi.prototype.loadUrl;
+  EditorUi.prototype.loadUrl = function(url, success, error, forceBinary, retry, dataUriPrefix) {
+    let exportedUrl = diagramUtils.getParameterValueFromURL('url', url);
+    if (exportedUrl) {
+      let diagramXML = diagramUtils.getDiagramXMLFromURL(exportedUrl);
+      if (diagramXML) {
+        success(diagramXML);
+      }
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramEditSheet.xml]: If ``diagramXML`` is null (can't be extracted from the URL) then the callback (success) is not called, which means the diagram editor is left waiting. We need to either call ``success`` or ``error`` ourselves or fallback on the original behavior (``originalLoadUrl``). The later option (``originalLoadUrl``) is probably the safest.

File: src/main/resources/Diagram/DiagramSheet.xml [hunk 15]
```diff
@@ -48,4 +48,141 @@
...
+    <property>
+      <code>//
+// Used in view &amp; edit sheets
+//
+define ('diagram-utils', ['mxgraph-common'], function() {
+  //
+  // Get a XML diagram from hash.
+  // The hash begins with 'R' &amp; continues with the encoded diagram.
+  //
+  function getDiagramXMLFromURL(url) {;
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramSheet.xml]:
```suggestion
  function getDiagramXMLFromURL(url) {
```

File: src/main/resources/Diagram/DiagramSheet.xml [hunk 16]
```diff
@@ -48,4 +48,141 @@
...
+
+  var urlParam = function(parameter, url) {
+    var results = new RegExp('[\?&amp;]' + parameter + '=([^&amp;#]*)').exec(url);
+    if (results === null) {
+       return null;
+    }
+    return decodeURIComponent(results[1]);
+  }
+
+  function getParameterValueFromURL(parameter, url) {
```
Comment (mflorea) [inline src/main/resources/Diagram/DiagramSheet.xml]: Inconsistent styling regarding function definition: either use

    var xxx = function(...) {...}

**everywhere**, or

    function xxx(...) {...}

My personal preference goes towards the first option (using ``var``). But what's important is to have a consistent code style.
